Implement config versioning

ekatef opened the pull request:
## Changes proposed in this Pull Request

Implements a check to ensure that the current local `config.yaml` is consistent with the version defined in `config.default.yaml`.


## Checklist

- [x] I consent to the release of this PR's code under the AGPLv3 license and non-code contributions under CC0-1.0 and CC-BY-4.0.
- [x] I tested my contribution locally and it seems to work fine.
- [ ] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `doc/requirements.txt`.
- [x] Changes in configuration options are added in all of `config.default.yaml` and `config.tutorial.yaml`.
- [ ] Add a test config or line additions to `test/` (note tests are changing the config.tutorial.yaml)
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

ekatef commented:
@davide-f have drafted a possible implementation. Could you please have a look? 

Have added one more field to the version number to track the changes in the config. Not sure it's the most elegant solution, though.

A possible improvement could be to add a check on the differences between the local config and `config.default.yaml`. Sometimes, it feels a good idea to have an opportunity to review the recent updates in the configuration options. However, not sure Snakemake traceback is the best place to place results of such a check. Probably, having an additional helper function would be a better idea. What is your feeling about that?

davide-f commented on Snakefile:
```diff
@@ -3,6 +3,7 @@
 # SPDX-License-Identifier: AGPL-3.0-or-later
 
 import sys
+import yaml
```
Great draft!

I'd propose to create an auxiliary function into helpers and place there all the code.
The function may by like: check_version(config)
Internally it loads the config.default and then does the magic.
What do you think?

ekatef replied:
Agree, revised.

davide-f commented on config.default.yaml:
```diff
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: CC0-1.0
 
-version: 0.3.0
+version: 0.3.0.0
```
Please revert :)

ekatef replied:
Done!

davide-f commented on config.tutorial.yaml:
```diff
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: CC0-1.0
 
-version: 0.3.0
+version: 0.3.0.0
```
Please revert

ekatef replied:
Done :)

davide-f reviewed (commented):
Great @ekatef :D
Beyond the minor comments, please add a release doc :)

ekatef commented:
> Great @ekatef :D Beyond the minor comments, please add a release doc :)

Done 🙂 Thanks a lot for the review!

I'm thinking what would be the best way to remind that a release is needed every time, when the config structure gets changed. Adding a point for that to PR template sounds like not a really good idea, but I don't have better ones at the moment. What is your perspective on that?

davide-f commented on Snakefile:
```diff
@@ -3,6 +3,7 @@
 # SPDX-License-Identifier: AGPL-3.0-or-later
 
 import sys
+import yaml
```
Can we drop this line? it seems is not needed

ekatef replied:
True, have forgotten to remove it. Thanks!

davide-f commented on scripts/_helpers.py:
```diff
@@ -27,6 +27,28 @@
 REGIONS_CONFIG = "regions_definition_config.yaml"
 
 
+def check_config_version(config):
+    """
+    Check that a version of the local config.yaml matches to the actual config
+    version as defined in config.default.yaml.
+    """
+
+    # using snakemake capabilities to deal with yanl configs
+    with open("config.default.yaml", "r") as f:
+        actual_config = yaml.safe_load(f)
+    actual_config_version = actual_config.get("version")
```
I'd propose to have the file path as parameter here, like:

```suggestion
def check_config_version(config, fp_config="config.default.yaml"):
    """
    Check that a version of the local config.yaml matches to the actual config
    version as defined in config.default.yaml.
    """

    # using snakemake capabilities to deal with yanl configs
    with open(fp_config, "r") as f:
        actual_config = yaml.safe_load(f)
    actual_config_version = actual_config.get("version")
```

ekatef replied:
Agree, done!

davide-f reviewed (commented):
Great @ekatef :D

I've added a minor comment. Moreover there is a minor merge conflict to fix.
This PR is ready to go afterwards. If you like to include these minor changes, that would be nice but the functionality is there.
Afterwards, if CI passes, feel free to merge yourself :)

ekatef commented:
> Great @ekatef :D
> 
> I've added a minor comment. Moreover there is a minor merge conflict to fix. This PR is ready to go afterwards. If you like to include these minor changes, that would be nice but the functionality is there. Afterwards, if CI passes, feel free to merge yourself :)

Fixed! Thank you 😄 Ok, then I'm merging it.

As a comment, the current implementation writes an error in logs in case the config is outdated, but don't terminate execution. Leaving that for now, while it can be improved further when introducing better management of the configs.
